Sources/SwiftKafka/Data/KafkaBuffer.swift

```diff
@@ -0,0 +1,25 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the swift-kafka-gsoc open source project
+//
+// Copyright (c) 2023 Apple Inc. and the swift-kafka-gsoc project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of swift-kafka-gsoc project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+/// A protocol representing a buffer that provides access to its raw memory representation.
+public protocol KafkaBuffer {
+    /// Yields a buffer pointer containing this `KafkaBuffer`'s readable bytes.
+    ///
+    /// - warning: Do not escape the pointer from the closure for later use.
+    ///
+    /// - parameters:
+    ///     - body: The closure that will accept the yielded bytes.
+    /// - returns: The value returned by `body`.
+    func withUnsafeRawBufferPointer<T>(_ body: (UnsafeRawBufferPointer) throws -> T) rethrows -> T
```

Comment:
Does this need to have such a low-level requirement? It feels like an API smell to me that the only requirement is an unsafe API.

Is it possible, for example, to rephrase the API in terms of e.g. `RandomAccessCollection` of `UInt8` and have encoders/decoders?

Reply:
I was thinking about this as well but I am unsure this fulfils the goals that we want to achieve here. We want to make it possible to have both `key` and especially `value` be able to pass copy-free to `rdkafka`. If we take a collection of UInt8 then I am not sure if we can achieve that.

Reply:
You do have `withContiguousStorageIfAvailable` for RAC which should fulfil that _most_ of the time (I guess only custom RACs _might_ not implement it).

Reply:
Yeah you are right we can probable be generic over `RAC` here instead of creating our own type. In the slow path we just write the `RAC` into byte buffer. @felixschlegel can you try this out.

Reply:
I agree that having the extra `KafkaBuffer` type is not ideal. However, I do not find the `RandomAccessCollection` approach more ergonomic (from what I have investigated):

1. Our `KafkaProducerMessage` type's `String` initializer would have to take `String` as parameter value, but stores a `ContiuousArray<UInt8>` which still requires copying

<details>
<summary>See implementation</summary>

```swift
/// Message that is sent by the `KafkaProducer`
public struct KafkaProducerMessage<Key: RandomAccessCollection<UInt8>, Value: RandomAccessCollection<UInt8>> {
    public var topic: String
    public var partition: KafkaPartition
    public var key: Key?
    public var value: Value

    /// Create a new `KafkaProducerMessage` with a key and value.
    /// - Parameter topic: The topic the message will be sent to. Topics may be created by the `KafkaProducer` if non-existent.
    /// - Parameter partition: The topic partition the message will be sent to. If not set explicitly, the partiotion will be assigned automatically.
    /// - Parameter key: Used to guarantee that messages with the same key will be sent to the same partittion so that their order is preserved.
    /// - Parameter value: The message body.
    public init(
        topic: String,
        partition: KafkaPartition? = nil,
        key: Key,
        value: Value
    ) {
        self.topic = topic
        self.key = key
        self.value = value

        if let partition = partition {
            self.partition = partition
        } else {
            self.partition = .unassigned
        }
    }
}

// MARK: - KafkaProducerMessage + String Initializer

extension KafkaProducerMessage where Key == ContiguousArray<UInt8>, Value == ContiguousArray<UInt8> {

    /// Create a new `KafkaProducerMessage` with a key and value.
    /// - Parameter topic: The topic the message will be sent to. Topics may be created by the `KafkaProducer` if non-existent.
    /// - Parameter partition: The topic partition the message will be sent to. If not set explicitly, the partiotion will be assigned automatically.
    /// - Parameter key: Used to guarantee that messages with the same key will be sent to the same partittion so that their order is preserved.
    /// - Parameter value: The message body.
    public init(
        topic: String,
        partition: KafkaPartition? = nil,
        key: String? = nil,
        value: String
    ) {
        self.topic = topic
        self.key = key.map { ContiguousArray($0.utf8) }
        self.value = ContiguousArray(value.utf8)

        if let partition = partition {
            self.partition = partition
        } else {
            self.partition = .unassigned
        }
    }
}
```

</details>

2. We would have to take `ByteBufferView` as parameter type instead of `ByteBuffer` as only `ByteBufferView` conforms to `RandomAccessCollection`

So yes it is a smell, but we provide default implementations of `KafkaBuffer` for `String` and `ByteBuffer`. The protocol gives the user the freedom to make their own types conform to `KafkaBuffer` at their own risk.  Furthermore this enables them to avoid copying their own types.

Reply:
Ah right, `UTF8View` is not a `RAC`, but it is a `BidirectionalCollection` which should be sufficient to avoid the copy. I don't think having convenience initialisers (for `String` and `ByteBuffer`) is an issue either. I'd rather reuse existing protocols if possible but don't feel too strongly.